k2/csrc/fsa_algo.cc

```diff
@@ -0,0 +1,141 @@
+
+// Copyright (c)  2020  Fangjun Kuang (csukuangfj@gmail.com)
+
+// See ../../LICENSE for clarification regarding multiple authors
+
+#include "k2/csrc/fsa_algo.h"
+
+#include <map>
+#include <stack>
+
+namespace {
+
+// depth first search state
+struct DfsState {
+  int32_t s;      // state number of the visiting node
+  int32_t begin;  // arc index of the visiting arc
+  int32_t end;    // end of the arc index of the visiting node
+};
+
+}  // namespace
+
+namespace k2 {
+
+// The implementation of this function is inspired by
+// http://www.openfst.org/doxygen/fst/html/connect_8h_source.html
+void ConnectCore(const Fsa &fsa, std::vector<int32_t> *state_map) {
+  // DCHECK_NOTNULL(state_map)
+  state_map->clear();
+  if (IsEmpty(fsa)) return;
+
+  auto num_states = fsa.NumStates();
+  auto final_state = num_states - 1;
+
+  std::vector<bool> accessible(num_states, false);
+  std::vector<bool> coaccessible(num_states, false);
+  std::vector<bool> visited(num_states, false);
+
+  accessible.front() = true;
+  coaccessible.back() = true;
+
+  std::stack<DfsState> stack;
+  stack.push({0, fsa.arc_indexes[0], fsa.arc_indexes[1]});
+  visited[0] = true;
+
+  while (!stack.empty()) {
+    auto &top = stack.top();
+
+    if (top.begin == top.end) {
+      // we have finished visiting this state
+      auto s = top.s;  // get a copy since we will destroy it
+      stack.pop();
+      if (!stack.empty()) {
+        // if it has a parent, set the parent's coaccessible flag
+        if (coaccessible[s]) {
+          auto &parent = stack.top();
+          coaccessible[parent.s] = true;
+          ++parent.begin;  // process the next child
+        }
+      }
+      continue;
+    }
+
+    const auto &arc = fsa.arcs[top.begin];
+    auto next_state = arc.dest_state;
+    bool is_visited = visited[next_state];
+    if (!is_visited) {
+      // this is a new discovered state
+      visited[next_state] = true;
+      auto begin = fsa.arc_indexes[next_state];
+      if (next_state != final_state)
+        stack.push({next_state, begin, fsa.arc_indexes[next_state + 1]});
+      else
+        stack.push({next_state, begin, begin});
+
+      if (accessible[top.s]) accessible[next_state] = true;
+    } else {
+      // this is a back arc or forward cross arc;
+      // update the coaccesible flag
+      auto next_state = arc.dest_state;
+      if (coaccessible[next_state]) coaccessible[top.s] = true;
+      ++top.begin;  // go to the next arc
+    }
+  }
+
+  state_map->reserve(num_states);
+
+  for (int32_t i = 0; i != num_states; ++i) {
+    if (accessible[i] && coaccessible[i]) state_map->push_back(i);
+  }
+}
+
+void Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
+  // DCHECK_NOTNULL(b);
+  std::vector<int32_t> state_map;
+  ConnectCore(a, &state_map);
+  if (state_map.empty()) return;
+
+  b->arc_indexes.resize(state_map.size());
+  b->arcs.clear();
+  b->arcs.reserve(a.arcs.size());
+
+  if (arc_map) {
+    arc_map->clear();
```

Comment:
delete?

Reply:
I do not have the information about the size of arc_map at this point, so I cannot use resize.

Reply:
In your pullrequest, you can remove the two variables: num state vistited and num arcs visited.
Then you can also use `reserve` instead of `resize`. 

Two fewer variables will be used.

Reply:
In my pullrequest, actually if I removed `num_visited_states`, I need to get the size of  `state_map_b2a` again and again in the loop which is inefficient. Furthermore, If I remove `num_visited_arcs`, as there may be cycle and duplicated visited arcs, I'll need to iterate over `visited_arcs` first to get `num_visited_arcs`.